### PR TITLE
Migrated EofParser core functionality from External Tool

### PR DESCRIPTION
**Pull request by DemuirGos**
Fixes Closes Resolves #

_Please choose one of the keywords above to refer to the issue this PR solves followed by the issue number (e.g. Fixes #000). If no issue number, remove the line. Also, remove everything marked optional that is not applicable. Remove this note after reading._

## Changes

* Adds an EofCodeParser/Validator 

#### What types of changes does your code introduce?

- [ ] Bugfix (a non-breaking change that fixes an issue)
- [ ] New feature (a non-breaking change that adds functionality)
- [ ] Breaking change (a fix or a feature that would cause existing functionality not to work as expected)
- [ ] Documentation update
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional or API changes)
- [ ] Build-related changes
- [x] Other: adds external tool

#### Requires testing

- [ ] Yes
- [x] No

#### If yes, did you write tests?

- [ ] Yes
- [x] No

#### Notes on testing

_Optional. Remove if not applicable._

## Remarks

_Optional. Remove if not applicable._

### CODE REVIEW

File: src/Nethermind/Nethermind.EofParser/LogInterceptor.cs [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+        }
+    }
+
+    void ILogger.Debug(string text) => throw new NotImplementedException();
+
+    void ILogger.Error(string text, Exception ex) => throw new NotImplementedException();
+
+    void ILogger.Info(string text) => throw new NotImplementedException();
+
+    void ILogger.Warn(string text) => throw new NotImplementedException();
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.EofParser/LogInterceptor.cs:38]: would not leave those as exceptions

File: src/Nethermind/Nethermind.EofParser/Program.cs [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+}
+
+void HandleLine(int idx, string line)
+{
+    if (line.StartsWith("#"))
+    {
+        return;
+    }
+
+    line = new string(line.Where(c => char.IsLetterOrDigit(c)).ToArray());
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.EofParser/Program.cs:25]: what other characters do we expect in input?

Comment (DemuirGos) [reply src/Nethermind/Nethermind.EofParser/Program.cs:25]: There was a discussion where they decided to add special characters to separate sections..., it was easier to just remove all unaccepted chars in a bytecode

File: src/Nethermind/Nethermind.EofParser/LogInterceptor.cs (hunk 1 above)
Comment (DemuirGos) [reply src/Nethermind/Nethermind.EofParser/LogInterceptor.cs:38]: Even if they won't be used?

Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.EofParser/LogInterceptor.cs:38]: They are not used now, but they can be used later - if you leave it at that you break `Liskov substitution principle`.

File: src/Nethermind/Nethermind.EofParser/Program.cs (hunk 2 above)
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.EofParser/Program.cs:25]: Can you add a comment for that?

Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.EofParser/Program.cs:25]: What if it has comments, starting with `#` but after some bytecode?

File: src/Nethermind/Nethermind.Evm/EOF/EvmObjectFormat.cs [hunk 3]
```diff
@@ -296,17 +298,8 @@ public bool ValidateBody(ReadOnlySpan<byte> container, in EofHeader header)
                 SectionHeader sectionHeader = header.CodeSections[sectionIdx];
                 (int codeSectionStartOffset, int codeSectionSize) = sectionHeader;
                 ReadOnlySpan<byte> code = container.Slice(codeSectionStartOffset, codeSectionSize);
-                if (!ValidateInstructions(code, header))
-                {
-                    if (Logger.IsTrace) Logger.Trace($"EIP-3670 : CodeSection {sectionIdx} contains invalid body");
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Evm/EOF/EvmObjectFormat.cs:301]: why remove this logs?

File: src/Nethermind/Nethermind.EofParser/LogInterceptor.cs [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+public static class LogSink
+{
+    public static StringBuilder ContentBuilder { get; set; } = new StringBuilder();
+    public static string Content => ContentBuilder.ToString();
+    public static void Clear()
+    {
+        LogSink.ContentBuilder.Clear();
+    }
+}
+public class InMemoryLogger : ILogger
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.EofParser/LogInterceptor.cs:15]: classes != filename

**Review by LukaszRozmej — APPROVED** [review]
general approve but please do look into those minor things

**Review by LukaszRozmej — COMMENTED** [review]
One more thing - where it should go, we were discussing some kind of `tools` space in the repo. @rubo ?

File: src/Nethermind/Nethermind.EofParser/Program.cs (hunk 2 above)
Comment (DemuirGos) [reply src/Nethermind/Nethermind.EofParser/Program.cs:25]: As per agreement comments aren't mixed with bytecodes in line so each line is either a comment | bytecode

**Comment by rubo** [comment]
> One more thing - where it should go, we were discussing some kind of tools space in the repo. @rubo ?

I'd say under `src/tools/`

File: src/Nethermind/Nethermind.Evm/EOF/EvmObjectFormat.cs [hunk 5]
```diff
@@ -9,6 +9,8 @@
 using Nethermind.Core.Extensions;
 using Nethermind.Logging;
 
+[assembly: InternalsVisibleTo("Nethermind.EofParser")]
+
 namespace Nethermind.Evm.EOF;
```
Comment (rubo) [inline src/Nethermind/Nethermind.Evm/EOF/EvmObjectFormat.cs:14]: Could you please change `EOF` to `Eof` to follow the .NET naming convention for 3+ letter shorthands? Both the namespace and directory.

**Review by rubo — CHANGES_REQUESTED** [review]

File: src/Nethermind/Nethermind.Evm/EOF/EvmObjectFormat.cs (hunk 5 above)
Comment (DemuirGos) [reply src/Nethermind/Nethermind.Evm/EOF/EvmObjectFormat.cs:14]: I think this review should be in this PR : https://github.com/NethermindEth/nethermind/pull/4950

File: src/Nethermind/Nethermind.Evm/EOF/EvmObjectFormat.cs (hunk 3 above)
Comment (DemuirGos) [reply src/Nethermind/Nethermind.Evm/EOF/EvmObjectFormat.cs:301]: they are redundant here, `ValidateInstructions` and `ValidateStackState` already log failure reason

**Pull request closed without merging**